### DISCUSSION TITLE
Qtree level user/group info

### DISCUSSION THREAD

**Discussion by jmg011** [NetApp/harvest#973] [Q&A] [answered]
Hi, 

Looking for a qtree level user and group info with this api: **qtree-list-iter**

The attrs do not have user/group info. Can you point me to an API that has this info at a qtree level?

building tree for attribute [attributes-list] => [qtree-info]

[qtree-info]                          -                                   *
  [export-policy]                     -                                   *
  [id]                                -                                   *
  [is-export-policy-inherited]        -                                   *
  [mode]                              -                                   *
  [oplocks]                           -                                   *
  [qos-policy-group]                  -                                   *
  [qtree]                             -                                   *
  [security-style]                    -                                   *
  [status]                            -                                   *
  [volume]                            -                                   *
  [vserver]                           -                                   *

**Comment by cgrinds** [comment, marked as answer]
hi @jmg011 user and group information come from the `quota-report-iter` ZAPI that the [qtree.yaml](https://github.com/NetApp/harvest/blob/ee5ecff977c95f137cadcaea2e0f6359de8b2cbb/conf/zapi/cdot/9.8.0/qtree.yaml#L20) template calls as part of the `Qtree` plugin.

I don't think the existing `Qtree` plugin collects user/group info and will require a code change to add. I'll check.

**Reply by cgrinds** [reply]
hey @jmg011 we discussed this a bit more and have some questions. Can you share more about your use case? 

`quota-report-iter` has the information you want, but there are a number of ways you may want to show the user/group information. One complication is a quota information can be expensive to gather because there may be many quotas with fan-out. The other issue is many users may be associated with a single qtree. In that case, would you want a comma separated list of labels or something else?

The qtree plugin, we added last release, filters on `quota-type=tree` to make the ZAPI call faster/less load. Depending on what you need, we might need to remove that filter.

**Reply by jmg011** [reply]
@cgrinds  Thank you for a quick response. I'm trying to get the owner and user set per qtree and not all the users using the qtree. 

[-user <user name>] - User ID
This optionally specifies the name or ID of the user that is set as the owner of the **qtree.**

[-group <group name>] - Group ID
This optionally specifies the name or ID of the group that is set as the owner of the **qtree.**

**Reply by cgrinds** [reply]
@jmg011 the qtree owner/group information is not available via ZAPI. Looks like it *is* available via the `/storage/qtrees` REST endpoint.

```
curl -k 'https://10.231.84.110/api/storage/qtrees?name=qt1&volume=avolume1&fields=user,group'
```

```json
{
  "records": [
    {
      "volume": {
        "uuid": "a941c6c3-e127-11ea-820e-00a098e0060e",
        "name": "avolume1",
        "_links": {
          "self": {
            "href": "/api/storage/volumes/a941c6c3-e127-11ea-820e-00a098e0060e"
          }
        }
      },
      "id": 1,
      "name": "qt1",
      "user": {
        "id": "991"
      },
      "group": {
        "id": "0"
      },
      "_links": {
        "self": {
          "href": "/api/storage/qtrees/a941c6c3-e127-11ea-820e-00a098e0060e/1"
        }
      }
    }
  ],
  "num_records": 1,
  "_links": {
    "self": {
      "href": "/api/storage/qtrees?name=qt1&volume=avolume1&fields=user,group"
    }
  }
}
```

**Comment by cgrinds** [comment]
Thanks to @Hardikl for tracking down the commands - pasted below in case we want to recreate later.

`cluster ocum-infinity`

```
qtree create -vserver test -volume avolume1 -qtree qt1 -security-style unix -oplock-mode enable -unix-permissions ---rwxr-xr-x -user 991

qtree show -vserver test -volume avolume1

Vserver    Volume        Qtree        Style        Oplocks   Status
---------- ------------- ------------ ------------ --------- --------
test       avolume1      ""           unix         enable    normal
test       avolume1      qt1          unix         enable    normal
2 entries were displayed.

qtree show -vserver test -volume avolume1 -qtree qt1
                                 Vserver Name: test
                                  Volume Name: avolume1
                                   Qtree Name: qt1
             Actual (Non-Junction) Qtree Path: /vol/avolume1/qt1
                               Security Style: unix
                                  Oplock Mode: enable
                                      User ID: 991
                                     Group ID: root
                             Unix Permissions: ---rwxr-xr-x
                                     Qtree Id: 1
                                 Qtree Status: normal
                                Export Policy: default
                   Is Export Policy Inherited: true
                             QoS policy group: -

```

**Reply by jmg011** [reply]
Thank you @cgrinds  @Hardikl  for confirming and providing details.
